### PR TITLE
feat: endless_runner template

### PR DESCRIPTION
**Pull request by spydon**
Adds the Endless Runner template to the repository.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by domesticmouse — APPROVED** [review]
LGTM once @filiph LGTM's

**Comment by domesticmouse** [comment]
@filiph I'd suggest landing this PR before #2, just so that the CI has something to, you know, CI.

File: templates/endless_runner/README.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+analytics, crash reporting, and so on, there are resources ready for you
+at [flutter.dev/games](https://flutter.dev/games).
+
+
+# Getting started
+
+Clone this project and run the following command in its root directory:
+
+```terminal
+flutter create . --project-name game_template
```
Comment (filiph) [inline templates/endless_runner/README.md:33]:
```suggestion
flutter create . --project-name endless_runner
```

File: templates/endless_runner/README.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+before you start any of the deeper integrations.
+[StackOverflow has instructions](https://stackoverflow.com/a/51550358/1416886)
+for this, and the [`rename`](https://pub.dev/packages/rename) tool
+(on pub.dev) automates the process.
+
+
+## Audio
+
+Audio is enabled by default and ready to go. You can modify code
+in `lib/src/audio/` to your liking.
```
Comment (filiph) [inline templates/endless_runner/README.md:146]:
```suggestion
in `lib/audio/` to your liking.
```

File: templates/endless_runner/README.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+## Settings
+
+The settings page is enabled by default, and accessible both
+from the main menu and through the "gear" button in the play session screen.
+
+Settings are saved to local storage using the
+[`shared_preferences`](https://pub.dev/packages/shared_preferences)
+package.
+To change what preferences are saved and how, edit files in
+`lib/src/settings/persistence`.
```
Comment (filiph) [inline templates/endless_runner/README.md:203]:
```suggestion
`lib/settings/persistence`.
```

Comment (filiph) [reply templates/endless_runner/README.md:203]: There may be more changes like this, but lets land first and look for these details in a follow-up.

File: templates/endless_runner/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    - assets/images/scenery/
+    - assets/music/
+    - assets/sfx/
+
+  fonts:
+    - family: Press Start 2P
+      fonts:
+        - asset: assets/fonts/Press_Start_2P/PressStart2P-Regular.ttf
+
+flutter_icons:
```
Comment (filiph) [inline templates/endless_runner/pubspec.yaml:49]:
```suggestion
flutter_launcher_icons:
```

This was a last-minute fix in the base template.

**Review by filiph — APPROVED** [review]
LGTM with a couple of suggestions.

After discussion with @domesticmouse, I decided to include platform subdirectories (I'm including `android/`, `ios/` and `web/`). It does have its advantages, plus it looks like the CI is currently not set up to create these subdirectories on its own.

**Comment by spydon** [comment]
> LGTM with a couple of suggestions.

Thanks for the fixes!

> It does have its advantages, plus it looks like the CI is currently not set up to create these subdirectories on its own.

What are the advantages of this? From what I have seen previously it usually just becomes more maintenance work to include the platform directories, since we don't do any changes in there and since they change between Flutter versions you want the users to have the latest platform directories created by `flutter create`, especially cumbersome when it comes to iOS and MacOS.

**Comment by filiph** [comment]
Oh, I just realized that there's no test in the template. Do you think you could add a simple smoke test, @spydon? It can be something similar to what's in `basic` or `card`. We want to ensure that the projects not only analyze well, but also run. It only catches the most egregious problems, but even _that_ is something.

**Comment by spydon** [comment]
> Oh, I just realized that there's no test in the template. Do you think you could add a simple smoke test, @spydon? It can be something similar to what's in `basic` or `card`. We want to ensure that the projects not only analyze well, but also run. It only catches the most egregious problems, but even _that_ is something.

Ah yes, I had that on my todo-list, but forgot about it!

**Comment by domesticmouse** [comment]
@filiph Ahh look, we now have CI on this PR, progress!

**Comment by domesticmouse** [comment]
Added `templates/endless_runner` to CI config, so this PR becomes actually self testing =)

**Comment by filiph** [comment]
Hi @spydon,

There are some analyzer problems:

```text
   info • The imported package 'logging' isn't a dependency of the importing package •
          lib/audio/audio_controller.dart:8:8 • depend_on_referenced_packages
warning • Unused import: 'package:flame/events.dart' • lib/flame_game/components/player.dart:3:8 •
       unused_import
   info • The imported package 'logging' isn't a dependency of the importing package •
          lib/settings/settings.dart:6:8 • depend_on_referenced_packages
```

**Comment by filiph** [comment]
> @filiph I'm guessing you are going to add a fix to this PR for the above issues?

I thought it would be automatic but it turns out to be a pull request on @spydon's repo. ¯\\_(ツ)\_/¯
https://github.com/spydon/games/pull/1

**Comment by domesticmouse** [comment]
> > @filiph I'm guessing you are going to add a fix to this PR for the above issues?
> 
> I thought it would be automatic but it turns out to be a pull request on @spydon's repo. ¯\_(ツ)_/¯ [spydon#1](https://github.com/spydon/games/pull/1)

hold my beer...

**Comment by filiph** [comment]
I know this is not the place to ask, but... how? How did you do this? Are you using the `gh` CLI tool?

**Comment by domesticmouse** [comment]
> I know this is not the place to ask, but... how? How did you do this? Are you using the `gh` CLI tool?

I'm using the [GitHub Desktop](https://desktop.github.com/) for mac to check out the repos, and my VSCode is logged into GitHub. Between all of the above it just worked ^(TM)

**Comment by domesticmouse** [comment]
Argh, this one we need to fix:

```
info - lib/main.dart:55:15 - 'useMaterial3' is deprecated and shouldn't be used. Use a ThemeData constructor (.from, .light, or .dark) instead. These constructors all have a useMaterial3 argument, and they set appropriate default values based on its value. See the useMaterial3 API documentation for full details. This feature was deprecated after v3.13.0-0.2.pre. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use
```

**Comment by spydon** [comment]
> I know this is not the place to ask, but... how? How did you do this? Are you using the `gh` CLI tool?

You don't need any special tool for that, as long as the PR owner has "Allow edits and access to secrets by maintainers" the maintainer can just check out the same branch as the PR owner and push directly to that one. It makes it simpler with the `gh` tool though, since you don't have to fiddle with remotes.

**Comment by domesticmouse** [comment]
@spydon it looks like your `flutterNesTheme` is not compliant with Flutter 3.16 beta. Can you go to `flutter channel beta` and figure out a fix? Alternatively, we can drop `useMaterial3: true` because I think that becomes default in Flutter 3.16. I think.

**Comment by spydon** [comment]
> @spydon it looks like your `flutterNesTheme` is not compliant with Flutter 3.16 beta. Can you go to `flutter channel beta` and figure out a fix? Alternatively, we can drop `useMaterial3: true` because I think that becomes default in Flutter 3.16. I think.

When is 3.16 to be released? If it is soon I suggest we just drop it.

**Comment by domesticmouse** [comment]
> When is 3.16 to be released? If it is soon I suggest we just drop it.

About the same time as this repo goes live, give or take.

**Comment by domesticmouse** [comment]
@filiph please feel free to land when you are happy. I'm going to bed soon =)

**Comment by spydon** [comment]
> Oh, I just realized that there's no test in the template. Do you think you could add a simple smoke test, @spydon? It can be something similar to what's in `basic` or `card`. We want to ensure that the projects not only analyze well, but also run. It only catches the most egregious problems, but even _that_ is something.

Added two simple smoke tests in https://github.com/flutter/games/pull/3/commits/e3ab2518e281d7d662f55987f6c43ad8818bc1a6

**Comment by filiph** [comment]
@spydon Looks like the flame test times out after 30seconds.

**Comment by filiph** [comment]
I've never seen this before. If you can't find a solution till tomorrow midday, I suggest you remove that Flame test for now so we can land this.

**Comment by spydon** [comment]
> I've never seen this before. If you can't find a solution till tomorrow midday, I suggest you remove that Flame test for now so we can land this.

There should be a work-around in place now.

**Comment by domesticmouse** [comment]
Landing on CI green for `stable` and `beta` for `ubuntu-latest` and `windows-latest`.  I don't see anything that should trigger in `macos-latest`, and those tests always draaaag.

**Pull request merged**